src/otk/external.py

```diff
@@ -1,85 +1,73 @@
 """Support for calling external programs to transform trees. External programs
-receive the current context and the subtree to operate on in JSON. They are
-expected to return the context and subtree.
-
-Returned subtrees are resolved during tree transformation."""
+receive the subtree to operate on in JSON. They are expected to return a new
+subtree."""
 
 import json
 import logging
+import pathlib
 import subprocess
 import sys
+import os
 from typing import Any
 
-from .context import Context, OSBuildContext
+from .context import Context
+from .constant import PREFIX_EXTERNAL
 
 log = logging.getLogger(__name__)
 
 
 def call(context: Context, directive: str, tree: Any) -> Any:
-    if directive not in registry:
-        log.fatal("unknown external %r", directive)
-
-        # TODO exception
-        sys.exit(1)
-        return
-
-    required_context, executable = registry[directive]
-
-    if not isinstance(context, required_context):
-        log.fatal("external %r with wrong context %r", directive, context)
-
-        # TODO exception
-        sys.exit(1)
-        return
+    exe, args = convert(directive)
+    exe = search(exe)
 
     data = json.dumps(
         {
-            "context": context.for_external(),
             "tree": tree,
         }
     )
 
     process = subprocess.run(
-        executable, input=data, encoding="utf8", capture_output=True
+        [exe] + args, input=data, encoding="utf8", capture_output=True
     )
 
     if process.returncode != 0:
         # TODO exception
         log.error(
             "call: %r %r %r failed: %r, %r",
-            executable,
-            context,
+            exe,
+            args,
             directive,
             process.stdout,
             process.stderr,
         )
         sys.exit(1)
         return
 
-    # Otherwise everything is good and we restore
-    data = json.loads(process.stdout)
+    return data["tree"]
 
-    context.from_external(data["context"])
 
-    # TODO restore context
-    return data["tree"]
+def convert(directive):
+    exe, *args = directive.removeprefix(PREFIX_EXTERNAL).split(".")
+    return (f"otk_external_{exe}", args)
+
+
+def search(exe):
+    paths = [
+        "/usr/local/libexec/otk",
+        "/usr/libexec/otk",
+        "/usr/local/lib/otk",
+        "/usr/lib/otk",
+    ]
+
+    env = os.getenv("OTK_EXTERNAL_PATH", None)
```

Comment:
I wonder if `OTK_EXTERNAL_PATH` should support `:`? But I guess we can always add it later if needed :)